src/nbl/ext/DebugDraw/CDrawAABB.cpp

```diff
@@ -0,0 +1,429 @@
+		std::make_pair(binding.memory->getCurrentMappingAccess().hasFlags(getRequiredAccessFlags(binding.memory->getMemoryPropertyFlags())), "Streaming buffer's memory current mapping access flags don't meet requirements!")
+		});
+
+	for (const auto& [ok, error] : validation)
+		if (!ok)
+		{
+			params.utilities->getLogger()->log(error, ILogger::ELL_ERROR);
+			return false;
+		}
+
+	return true;
+}
+
+smart_refctd_ptr<IGPUBuffer> DrawAABB::createIndicesBuffer(SCreationParameters& params)
+{
+	std::array<uint32_t, IndicesCount> unitAABBIndices;
+	unitAABBIndices[0] = 0;
+	unitAABBIndices[1] = 1;
+	unitAABBIndices[2] = 0;
+	unitAABBIndices[3] = 2;
+
+	unitAABBIndices[4] = 3;
+	unitAABBIndices[5] = 1;
+	unitAABBIndices[6] = 3;
+	unitAABBIndices[7] = 2;
+
+	unitAABBIndices[8] = 4;
+	unitAABBIndices[9] = 5;
+	unitAABBIndices[10] = 4;
+	unitAABBIndices[11] = 6;
+
+	unitAABBIndices[12] = 7;
+	unitAABBIndices[13] = 5;
+	unitAABBIndices[14] = 7;
+	unitAABBIndices[15] = 6;
+
+	unitAABBIndices[16] = 0;
+	unitAABBIndices[17] = 4;
+	unitAABBIndices[18] = 1;
+	unitAABBIndices[19] = 5;
+
+	unitAABBIndices[20] = 2;
+	unitAABBIndices[21] = 6;
+	unitAABBIndices[22] = 3;
+	unitAABBIndices[23] = 7;
+
+	IGPUBuffer::SCreationParams bufparams;
+	bufparams.size = sizeof(uint32_t) * unitAABBIndices.size();
+	bufparams.usage = IGPUBuffer::EUF_INDEX_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+
+	smart_refctd_ptr<IGPUBuffer> indicesBuffer;
+	params.utilities->createFilledDeviceLocalBufferOnDedMem(
+		SIntendedSubmitInfo{ .queue = params.transfer },
+		std::move(bufparams),
+		unitAABBIndices.data()
+	).move_into(indicesBuffer);
+
+	return indicesBuffer;
+}
+
+core::smart_refctd_ptr<video::IGPUPipelineLayout> DrawAABB::createPipelineLayoutFromPCRange(video::ILogicalDevice* device, const asset::SPushConstantRange& pcRange)
+{
+	return device->createPipelineLayout({ &pcRange , 1 }, nullptr, nullptr, nullptr, nullptr);
+}
+
+core::smart_refctd_ptr<video::IGPUPipelineLayout> DrawAABB::createDefaultPipelineLayout(video::ILogicalDevice* device)
+{
+	SPushConstantRange pcRange = {
+		.stageFlags = IShader::E_SHADER_STAGE::ESS_VERTEX,
+		.offset = 0,
+		.size = sizeof(SPushConstants)
+	};
+	return device->createPipelineLayout({ &pcRange , 1 }, nullptr, nullptr, nullptr, nullptr);
+}
+
+bool DrawAABB::renderSingle(IGPUCommandBuffer* commandBuffer, const hlsl::shapes::AABB<3, float>& aabb, const hlsl::float32_t4& color, const hlsl::float32_t4x4& cameraMat)
+{
+	if (!(m_cachedCreationParams.drawMode & ADM_DRAW_SINGLE))
+	{
+		m_cachedCreationParams.utilities->getLogger()->log("DrawAABB has not been enabled for draw single!", ILogger::ELL_ERROR);
+		return false;
+	}
+
+	commandBuffer->bindGraphicsPipeline(m_singlePipeline.get());
+	commandBuffer->setLineWidth(1.f);
```

Comment:
make a struct with command buffer pointer, camera viewProj matrix and line width to use for both functions